### ISSUE TITLE
[Bug]: projectService failed to find referenced tsconfigs for typical monorepo setup

### ISSUE THREAD

**Issue by swwind** [web-infra-dev/rslint#1234]
### Details
There is a mismatched behavior of the `projectService` configuration option between `rslint` and `eslint`.

In a monorepo setup where the root `tsconfig.json` only contains `references` pointing to package-specific config files, `rslint` (configured with `projectService: true`) fails to find the corresponding `tsconfig.json` for the TypeScript file, causing type-aware rules to be silently skipped.

In contrast, `eslint` (configured with `projectService: true`) successfully follows the references, locates the correct `tsconfig.json` for the file, and runs type-aware checks correctly.

For more details on the setup and comparison, see #1225.

### Reproduce link
https://github.com/swwind/rslint-reference-repro

### Reproduce Steps
1. Clone the repro repository: `https://github.com/swwind/rslint-reference-repro`
2. Run `rslint` with the configuration in [rslint3.config.ts](file:///home/swwind/ts-monorepo/rslint3.config.ts) (which has `parserOptions: { projectService: true }`):
   ```bash
   pnpm exec rslint --config rslint3.config.ts .
   ```
   Observe that it only reports the `no-console` error but silently skips the `@typescript-eslint/no-floating-promises` error on [index.ts](file:///home/swwind/ts-monorepo/index.ts) (reports 1 error instead of 2).
3. Run `eslint` with the configuration in [eslint4.config.ts](file:///home/swwind/ts-monorepo/eslint4.config.ts) (which also has `parserOptions: { projectService: true }`):
   ```bash
   pnpm exec eslint -c eslint4.config.ts .
   ```
   Observe that it successfully reports 2 errors (both `no-console` and `@typescript-eslint/no-floating-promises`).
